lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -1764,9 +1764,27 @@ ParseResult FIRStmtParser::parseLeadingExpStmt(Value lhs, SubOpVector &subOps) {
       parseOptionalInfo(info, subOps))
     return failure();
 
-  if (kind == FIRToken::less_equal)
+  if (kind == FIRToken::less_equal) {
+    // Some operations, dshl for example, have implicit truncations, even in lo
+    // firrtl.  Handle truncations in a connect here.
+    if (lhs.getType() != rhs.getType()) {
+      auto lhsType =
+          lhs.getType().cast<FIRRTLType>().getPassiveType().dyn_cast<IntType>();
+      auto rhsType =
+          rhs.getType().cast<FIRRTLType>().getPassiveType().dyn_cast<IntType>();
+      if (lhsType && rhsType && lhsType.hasWidth() && rhsType.hasWidth() &&
+          lhsType.getWidthOrSentinel() < rhsType.getWidthOrSentinel()) {
+        rhs = builder.create<TailPrimOp>(
+            info.getLoc(),
+            IntType::get(builder.getContext(), false,
+                         lhsType.getWidthOrSentinel()),
+            rhs, lhsType.getWidthOrSentinel());
+        if (lhsType.isSigned()) // rhs is unsigned
+          rhs = builder.create<AsSIntPrimOp>(info.getLoc(), lhsType, rhs);
```

Comment:
Adding the cast opens up the set of circuits we accept. Specifically, the SFC wants to add the cast for the `dshl`, but I'm not sure about generally adding that. E.g., the SFC rejects the following, but this PR will accept it:

```
circuit Foo :
  module Foo :
    input a: UInt<2>
    output b: SInt<1>
    b <= a
```

```bash
firrtl -i Foo.fir
# Exception in thread "main" firrtl.passes.CheckTypes$InvalidConnect: : [module Foo]  Type mismatch in 'b <= a'.
#   b: SInt<1>
#   a: UInt<2>
```

```bash
firtool Foo.fir  
# module  {
#   firrtl.circuit "Foo" {
#     firrtl.module @Foo(%a: !firrtl.uint<2>, %b: !firrtl.flip<sint<1>>) {
#       %0 = firrtl.bits %a 0 to 0 : (!firrtl.uint<2>) -> !firrtl.uint<1>
#       %1 = firrtl.asSInt %0 : (!firrtl.uint<1>) -> !firrtl.sint<1>
#       firrtl.connect %b, %1 : !firrtl.flip<sint<1>>, !firrtl.sint<1>
#     }
#   }
# }
```

I'm looking into what's going on in the SFC here.

Note: this may be fine as it's still in the spirit of accepting what the SFC accepts, but we accept more.

Reply:
Correction to the above: the SFC is doing the correct thing here. There's no problem there.

This PR can only do the cast if the LHS and original RHS are SInts. Currently it's doing the cast unconditionally which is causing the promotion behavior from a (too wide) UInt to SInt.

Reply:
Yeah, I agree that this is a heavy hammer.  Making this as narrow as possible would be preferable.  The patch LGTM otherwise.

Did you look at fixing SFC to not emit the invalid code here, or is that impractical?

I'm happy for you two to work out what the minimal solution is for this, weighing the tradeoffs of the various approaches.

Reply:
You can also narrow this by checking to see if the RHS is a whitelisted set of ops like `dshl` if that is the only case where this occurs.

Reply:
I was thinking about turning this into a partial connect, but it seems like partial connects are suppose to be whole elements of structs or arrays, not some bits of a primitive type (that way lies madness!).  If we can I'd like to keep distinct and clear partial connect semantics, thus inserting the truncation is seeming the correct thing.

Reply:
Further clarification: the SFC is hamstrung because of what Chisel3 emits. Chisel3 will emit either of the two Chisel3 connect operators (`:=` or `<>`) as expanded, element-wise FIRRTL connects. E.g.:

```scala
class Foo extends RawModule {
  val a = IO(Input(UInt(2.W)))
  val b = IO(Output(UInt(1.W)))
  b := a 
}
```

Gives you:

```
circuit Foo : 
  module Foo : 
    input a : UInt<2>
    output b : UInt<1>
    
    b <= a
```

Chisel devs have talked about changing this to be partial connect (which would avoid the need for this in CIRCT if consuming FIRRTL IR coming from a Chisel release after that). We do still need some solution to deal with legacy Chisel-produced FIRRTL, though...

The SFC could clean up this code internally, but it's likely not worth it.

Reply:
And the solution here LGTM if modified to cast only when it should. :+1:

Reply:
Ok, seems like we should do this, and I agree with you that partial connect seems best to keep to aggregates.  Please narrow this if possible to specific ops.

Reply:
So talking to @seldridge  offline, my understanding is that a partial connect of a primitive value can truncate, but a partial connect of a bundle will not do truncation of it's members.  Given this, I'll take this approach and leave partial connects to bundles only.  This is a deviation from FIRRTL and I'll update the rational document.

Reply:
Sorry, I may have said something wrong... Partial connect of an element or an aggregate will truncate in either case. E.g.:

```
circuit Bar :
  module Bar :
    input a: { a: UInt<8>, b: UInt<8>}
    output b: { a: UInt<4>, b: UInt<4>}
    b <- a
```

Produces (`firrtl -i Bar.fir`):

```verilog
module Bar(
  input  [7:0] a_a,
  input  [7:0] a_b,
  output [3:0] b_a,
  output [3:0] b_b
);
  assign b_a = a_a[3:0];
  assign b_b = a_b[3:0];
endmodule
```